### PR TITLE
Add CanonicalizeName function for kbfs TLFs

### PR DESCRIPTION
**Pull request by patrickxb**
r? @maxtaco

### CODE REVIEW

File: go/libkb/kbfs_folder.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+package libkb
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+)
+
+var caseSensitiveServices = map[string]bool{"hackernews": true}
```
Comment (maxtaco) [inline go/libkb/kbfs_folder.go]: I've made a habit of putting all HackerNews-related code into `proof_support_hackernews`.  This way when we add another service, we don't have stuff peppered all over the repo...

File: go/libkb/kbfs_folder.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+		// convert from service:name to name@service
+		if strings.Contains(id, ":") {
+			pieces := strings.Split(id, ":")
+			id = pieces[1] + "@" + pieces[0]
+		}
+
+		// some services have case-sensitive usernames
+		pieces := strings.Split(id, "@")
+		if len(pieces) == 2 && caseSensitiveServices[pieces[1]] {
```
Comment (maxtaco) [inline go/libkb/kbfs_folder.go]: KB names should be tolowered.  Also the service names should be tolowered, and before they're checked against the map above.  Someone can try `maxtaco@TwiTTEr`, we should fix that.

File: go/libkb/kbfs_folder_test.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	{in: "bob/", out: "", err: true},
+	{in: "bob@foo@bar", out: "", err: true},
+	{in: "foo:bar:bob", out: "", err: true},
+	{in: "foo:bob@bar", out: "", err: true},
+	{in: "charlie,bob,alice", out: "alice,bob,charlie"},
+	{in: "BOB@coinbase", out: "bob@coinbase"},
+	{in: "BOB@github", out: "bob@github"},
+	{in: "BOB@hackernews", out: "BOB@hackernews"},
+	{in: "BOB@reddit", out: "bob@reddit"},
+	{in: "BOB@rooter", out: "bob@rooter"},
```
Comment (maxtaco) [inline go/libkb/kbfs_folder_test.go]: also add tests for the cases mentioned above...

File: go/libkb/kbfs_folder.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+// canonical version of it. It also checks for basic errors.
+func CanonicalizeName(name string) (string, error) {
+	// check for invalid characters
+	if strings.ContainsAny(name, "/+") {
+		return "", errors.New("invalid character in name")
+	}
+
+	// split into individual ids
+	ids := strings.Split(name, ",")
+	for i, id := range ids {
```
Comment (maxtaco) [inline go/libkb/kbfs_folder.go]: probably want to complain about empty names.

Also what about `#`-style readers?

Comment (maxtaco) [inline go/libkb/kbfs_folder.go]: Maybe split the inside of this loop into its own function...

File: go/libkb/kbfs_folder.go [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+		// some services have case-sensitive usernames
+		pieces := strings.Split(id, "@")
+		if len(pieces) == 2 && caseSensitiveServices[pieces[1]] {
+			ids[i] = id
+		} else {
+			ids[i] = strings.ToLower(id)
+		}
+	}
+	sort.Strings(ids)
```
Comment (maxtaco) [inline go/libkb/kbfs_folder.go]: @oconnor663 did it this way, is this really the right idea?

`/keybase/max@twitter,norman` or `/keybase/norman,max@twitter`

I'm not sure...

Comment (oconnor663) [inline go/libkb/kbfs_folder.go]: Here's what we did on the web side, for comparison. Definitely quick and dirty: https://github.com/keybase/keybase/blob/master/web/kbfs.iced#L25

**Comment by strib** [comment]
Not sure if this is meant to be the only PR for CORE-2667, but it's not at all what I expected.  We don't need and won't use something for a full TLF name, and I think we should avoid putting kbfs-specific stuff in `client` anyway.  I thought 2667 was supposed to just be about canonicalizing single assertions.  KBFS does it's own parsing and ordering of TLF names.

File: go/libkb/kbfs_folder.go (hunk 5 above)
Comment (strib) [inline go/libkb/kbfs_folder.go]: We do this ourselves in KBFS.  We wouldn't use a `libkb` function for it.

Comment (strib) [inline go/libkb/kbfs_folder.go]: For the record, this is what we do now, without unresolved assertions: https://github.com/keybase/kbfs/blob/master/libkbfs/tlf_handle.go#L92-L109

I imagine we'll just sort the writer name and assertions all together alphabetically, but I don't really care too much.

**Comment by patrickxb** [comment]
@strib PTAL? This should work as we discussed.

**Comment by strib** [comment]
Looks good to me, thanks!  (Assuming the appveyor failure is from something else.)

**Pull request merged**